### PR TITLE
[GENFI] Update genfi patch

### PR DESCRIPTION
**Pull request by MatthieuJoulot**
Processing the whole dataset reveals a few more problems. This PR handles these problems.

### CODE REVIEW

**Comment by MatthieuJoulot** [comment]
The test should break since the CI needs to be updated

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 1]
```diff
@@ -795,10 +815,11 @@ def merge_philips_diffusion(
     multipart_id = _get_multipart_id(
         PhilipsNumberOfParts.from_int(int(number_of_parts)), run_num
     )
-    if multipart_id is not None:
-        data = json.loads(json_file.read_text())
-        data["MultipartID"] = multipart_id
-        json.dump(data, json_file, indent=4)
+    with open(json_file, "r+") as f:
+        if multipart_id is not None:
+            data = json.load(f)
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:820]: Why these changes ?

File: docs/Converters/GENFItoBIDS.md [hunk 2]
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
 - Fieldmaps
 - rsfMRI
 
+## Downloading GENFI
+
+To download GENFI in a way that you can convert it, you only need to make sure of two thing:
```
Comment (NicolasGensollen) [inline docs/Converters/GENFItoBIDS.md:22]:
```suggestion
To download GENFI in a way that you can convert it, you need to make sure of two things: 
```

File: docs/Converters/GENFItoBIDS.md [hunk 3]
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
 - Fieldmaps
 - rsfMRI
 
+## Downloading GENFI
+
+To download GENFI in a way that you can convert it, you only need to make sure of two thing: 
+- To only select the "simplify downloaded archive structure" in download data options.
```
Comment (NicolasGensollen) [inline docs/Converters/GENFItoBIDS.md:23]:
```suggestion
- Only select the "simplify downloaded archive structure" in download data options.
```

**Comment by MatthieuJoulot** [comment]
Unit-tests currently crash for the following reason: I changed the content of the `bidsignore` and added `dwi` to be ignored. I did that in order to respect the Philips sequence and to pass the bids-validator, which otherwise contradict.

They contradict because GENFI acquired a single b0, which has no bvecs or bvals, which is not allowed by the bids-spec.

I'm conflicted regarding this addition of dwi to the bidsignore, I'd rather remove it (even though I've added it), and not let GENFI pass the bids-validator, since it is out of spec.
@NicolasGensollen WDYT ?

File: docs/Converters/GENFItoBIDS.md [hunk 4]
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
...
+- And in Select Image Data section Scans Format, select only "DICOM"
```
Comment (NicolasGensollen) [inline docs/Converters/GENFItoBIDS.md:24]: I'm not sure my suggestion is correct, please verify where the `"` should be placed.

```suggestion
- Select only "DICOM" in "Select Image Data" section "Scans Format".
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @MatthieuJoulot !
I made a quick first pass without testing it (will do that later today...).
I'm a bit concerned with the changes made to utility functions used across all converters.
Did you make sure your modifications were not breaking anything ?

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 5]
```diff
@@ -640,15 +648,23 @@ def merge_imaging_data(df_dicom: DataFrame) -> DataFrame:
...
         bids_filename=lambda df: df[
-            ["participant_id", "session_id", "run_num", "suffix"]
+            ["participant_id", "session_id", "run_num", "task", "suffix"]
         ].agg("_".join, axis=1),
+        # bids_full_path=lambda df: df[
+        #     ["participant_id", "session_id", "datatype", "bids_filename"]
+        # ].agg("/".join, axis=1),
+    )
+    df_sub_ses_run = df_sub_ses_run.assign(
+        bids_filename=lambda df: df.bids_filename.apply(lambda x: x.replace("__", "_"))
```
[the remark this answers is not in the document]
Comment (MatthieuJoulot) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:660]: because of the task_rest that is aggregated as well. since it only exists for fMRI acquisitions, two underscores are aggregated for other modalities.

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py (hunk 1 above)
Comment (MatthieuJoulot) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:820]: I don't know why, but this doesn't work. So I changed back to what I had done in the first place, which I does !
It may not be optimal though, so we can look at it together

File: clinica/iotools/bids_utils.py [hunk 6]
```diff
@@ -905,6 +911,9 @@ def run_dcm2niix(
             ),
             lvl="warning",
         )
+        return False
+    else:
+        return True
```
Comment (NicolasGensollen) [inline clinica/iotools/bids_utils.py:916]:
```suggestion
    return True
```

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 7]
```diff
@@ -709,13 +721,17 @@ def write_bids(
...
+        dcm2niix_success = run_dcm2niix(
             Path(metadata["source_path"]).parent,
             to / str(Path(bids_full_path).parent),
             metadata["bids_filename"],
             True,
         )
-        if "dwi" in metadata["bids_filename"] and metadata.manufacturer == "Philips":
+        if (
+            "dwi" in metadata["bids_filename"]
+            and "Philips" in metadata.manufacturer
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:732]: Why the change from equality to a weaker inclusion here ?

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py (hunk 1 above)
Comment (NicolasGensollen) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:816]: OK, we'll take a look at the error you're getting this afternoon.

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 8]
```diff
@@ -835,7 +852,7 @@ def _get_multipart_id(nb_parts: PhilipsNumberOfParts, run_num: str) -> Optional[
             return "dwi_1"
         raise ValueError(f"{run_num} is outside of the scope.")
     if nb_parts == PhilipsNumberOfParts.FIVE:
-        if run_num in (f"run-0{k}" for k in range(1, 5)):
+        if run_num in (f"run-0{k}" for k in range(1, 6)):
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:855]: Was the 5 a mistake or has something changed ?

File: clinica/iotools/bids_utils.py [hunk 9]
```diff
@@ -885,6 +886,11 @@ def run_dcm2niix(
 
     bids_sidecar : bool, optional
         Whether to generate a BIDS sidecar or not. Default=True.
+
+    Returns
+    -------
+    dcm2niix_success: bool
```
Comment (NicolasGensollen) [inline clinica/iotools/bids_utils.py:892]: Since you're not naming the return value:

```suggestion
    bool :
```

**Review by NicolasGensollen — COMMENTED** [review]
Made a second pass. I think we need to refactor a bit the function `merge_imaging_data`. It's getting really difficult to understand what's going on in it and I'm afraid it will look totally opaque to us in a few months...

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py (hunk 8 above)
Comment (MatthieuJoulot) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:906]: No ! We want to have the 5 first runs

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py (hunk 7 above)
Comment (MatthieuJoulot) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:783]: Sometimes the manufacturer is not just `Philips`but `Philips &Co`or whatnot

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 10]
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
...
+    df = _compute_runs(df)
+    df = _compute_bids_full_path(df)
+
+    return df
+
+
+def _compute_source_id_and_source_ses_id(df: DataFrame) -> DataFrame:
+    """Adds two columns built from the column 'source'.
+
+    - 'source_id': subject ID in the raw dataset
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:299]: Could you add a line for `source` as well ?

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 11]
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
...
+    - 'source_ses_id': Session ID in the raw dataset
+
+    Example
+    -------
+    source_id = 'C9ORF001'
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:304]: Same, it'd be great to have the `source` column in your example

File: clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py [hunk 12]
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
...
+    source_ses_id = '11
```
Comment (NicolasGensollen) [inline clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:305]: Why 11 ?

Comment (NicolasGensollen) [reply clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:307]: It's much clearer now that you added an example for `source`, thanks !

**Review by NicolasGensollen — APPROVED** [review]
LGTM, thanks @MatthieuJoulot !

**Comment by NicolasGensollen** [comment]
Alright ! All tests are passing, merging...
